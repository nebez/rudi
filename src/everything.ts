export function injectable(): ClassDecorator {
    return () => {};
}

type Newable<T> = { new (...args: any[]): T };
type Abstraction<T> = Function & { prototype: T };

interface Scope {
    singletonScope(): void;
    resolutionScope(): void;
}

type FactoryFunction<T> = (container: Container) => T;

export interface Container {
    register<To>(to: Newable<To>): Scope;
    register<From, To extends From>(from: Abstraction<From>, to: Newable<To>): Scope;
    register<From, To extends From>(from: Abstraction<From>, to: To): void;
    register<From, To extends From>(from: Abstraction<From>, to: FactoryFunction<To>): Scope;
}
