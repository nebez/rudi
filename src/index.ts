export function injectable(): ClassDecorator { return () => {} }

type Newable<T> = { new (...args: any[]): T };
type Abstraction<T> = Function & { prototype: T };
type FactoryFunction<T> = (container: Container) => T;

interface Scope {
    singletonScope(): void;
    resolutionScope(): void;
}

interface AbstractRegistration<T, U extends T> {
    type: 'abstract';
    token: Abstraction<T>;
    class: Newable<U>;
}

interface InstanceRegistration<T, U extends T> {
    type: 'instance';
    token: Abstraction<T>;
    instance: U;
}

interface FactoryRegistration<T, U extends T> {
    type: 'factory';
    token: Abstraction<T>;
    factory: FactoryFunction<U>;
}

type ContainerRegistration<T, U extends T> =
    AbstractRegistration<T, U> |
    InstanceRegistration<T, U> |
    FactoryRegistration<T, U>;

export interface Container {
    register<To>(to: Newable<To>): Scope;
    register<From, To extends From>(from: Abstraction<From>, to: Newable<To>): Scope;
    // instance<From, To extends From>(from: Abstraction<From>, to: To): void;
    // factory<From, To extends From>(from: Abstraction<From>, to: FactoryFunction<To>): Scope;
    resolve<T>(token: Abstraction<T>): T;
}

export class FirstContainer implements Container {
    protected registrations = new Map<Abstraction<any>, ContainerRegistration<any, any>>();

    public register<To>(to: Newable<To>): Scope;
    public register<From, To extends From>(from: Abstraction<From>, to: Newable<To>): Scope;
    public register<From, To extends From>(...args: [Newable<To>] | [Abstraction<From>, Newable<To>]): Scope {
        let registration;

        if (args.length === 1) {
            registration = { token: args[0], class: args[0] };
        } else {
            registration = { token: args[0], class: args[1] };
        }

        if (registration.class.length >= 1 && this.getParams(registration.class) == undefined) {
            throw new Error(`Attempting to register ${registration.class.name} that has constructor args without an annotation`);
        }

        this.registrations.set(args[0], {
            type: 'abstract',
            ...registration,
        });

        return { singletonScope: () => {}, resolutionScope: () => {} }
    }

    public resolve<T>(token: Abstraction<T>): T {
        const registration = this.registrations.get(token);

        if (registration == undefined) {
            // todo - can we just construct it if it's newable and has no args?
            // return new (token as any)();
            throw new Error(`No registration found for ${token.name}. Did you forget to add it to the container?`);
        }

        if (registration.type === 'factory' || registration.type === 'instance') {
            throw new Error(`Not implemented yet`);
        }

        const args = this.createConstructorArguments(registration.class);

        return new registration.class(...args);
    }

    protected createConstructorArguments<T>(forClass: Newable<T>): any[] {
        const params = this.getParams(forClass);

        if (params == undefined) {
            // throw new Error('Missing something for ' + forClass.name);
            return [];
        }

        return params.map((x: Abstraction<T>) => this.resolve(x));
    }

    protected getParams<T>(forClass: Newable<T>): any[] | undefined {
        return Reflect.getMetadata('design:paramtypes', forClass);
    }
}
