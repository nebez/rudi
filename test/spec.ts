// This file is type-checked against the compiler. You can't execute it
// because we're testing our specification, not the implementation.
import { Container } from '../src/everything';

abstract class Logger {
    abstract log(message: string): void;
}

class ConsoleLogger {
    log(message: string) {
        console.log(message);
    }
}

class BadImplementationLogger {
    log(message: number) {
        console.log(message);
    }
}

class Car {
    drive() {
        console.log('driving!');
    }
}

let container: Container;

/**
 * These should be fine.
 */

// We can bind a concrete class alone, or a concrete class to another concrete
// class of the same shape. We can also bind abstractions to their concrete
// implementations.
container.register(ConsoleLogger);
container.register(ConsoleLogger, ConsoleLogger);

container.register(Car);
container.register(Car, Car);

container.register(Logger, ConsoleLogger);

// Self-binding and abstract-to-concrete bindings can have different
// scopes: singleton and request.
container.register(ConsoleLogger).singletonScope();
container.register(ConsoleLogger).resolutionScope();
container.register(ConsoleLogger, ConsoleLogger).singletonScope();
container.register(ConsoleLogger, ConsoleLogger).resolutionScope();

container.register(Car).singletonScope();
container.register(Car).resolutionScope();
container.register(Car, Car).singletonScope();
container.register(Car, Car).resolutionScope();

container.register(Logger, ConsoleLogger).singletonScope();
container.register(Logger, ConsoleLogger).resolutionScope();

// We should also enable instance registration.
const myLogger = { log: (message: string) => { console.error(message); } };
const myLogger2 = new ConsoleLogger();

container.register(Logger, myLogger);
container.register(Logger, myLogger2);

container.register(ConsoleLogger, myLogger);
container.register(ConsoleLogger, myLogger2);

// Factory registrations would be nice. Factory resolutions can be scoped and
// optionally have full access to the container if desired.
container.register(Logger, () => new ConsoleLogger());
container.register(ConsoleLogger, () => new ConsoleLogger());
container.register(Logger, (c: Container) => new ConsoleLogger());

container.register(ConsoleLogger, (c: Container) => {
    return { log: (message: string) => { console.error(message); } };
});

container.register(Logger, () => {
    if (Math.random() > 0.5) {
        return myLogger;
    }

    return myLogger2;
});

container.register(Logger, () => new ConsoleLogger()).singletonScope();
container.register(Logger, () => new ConsoleLogger()).resolutionScope();
container.register(ConsoleLogger, () => new ConsoleLogger()).singletonScope();
container.register(ConsoleLogger, () => new ConsoleLogger()).resolutionScope();
container.register(Logger, (c: Container) => new ConsoleLogger()).singletonScope();
container.register(Logger, (c: Container) => new ConsoleLogger()).resolutionScope();

/**
 * These should not be fine.
 */

// You can't bind a concrete class to another concretion of a different shape.
container.register(Car, ConsoleLogger);
container.register(ConsoleLogger, Car);

// You can't bind an abstract class to itself.
container.register(Logger);
container.register(Logger, Logger);

// You can't bind an abstraction to a concretion of the wrong shape.
container.register(Logger, Car);
container.register(Logger, BadImplementationLogger);

// You shouldn't be able to change the scope of an instance binding. These
// should all be errors.
container.register(Logger, myLogger).singletonScope();
container.register(Logger, myLogger).resolutionScope();
container.register(Logger, myLogger2).singletonScope();
container.register(Logger, myLogger2).resolutionScope();

// You can't attach a wrong shape to an abstraction.
const myBadLogger = { log: (message: string[]) => { console.error(message); } };
container.register(Logger, myBadLogger);
container.register(ConsoleLogger, myBadLogger);
container.register(Car, myBadLogger);

// Factory functions must return the right shape - always.
container.register(Logger, () => new Car());

container.register(Logger, () => {
    if (Math.random() > 0.5) {
        return myLogger;
    }

    return 'this shouldnt happen';
});


container.register(ConsoleLogger, () => {
    return { log: (message: string[]) => { console.error(message); } };
});

// Factory functions have access to the container and nothing else.
container.register(ConsoleLogger, (c: number) => new ConsoleLogger());
