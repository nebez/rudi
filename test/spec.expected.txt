test/spec.ts(99,25): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Car>, to: Newable<Car>): Scope', gave the following error.
    Argument of type 'typeof ConsoleLogger' is not assignable to parameter of type 'Newable<Car>'.
      Property 'drive' is missing in type 'ConsoleLogger' but required in type 'Car'.
  Overload 2 of 4, '(from: Abstraction<Car>, to: Car): void', gave the following error.
    Argument of type 'typeof ConsoleLogger' is not assignable to parameter of type 'Car'.
      Property 'drive' is missing in type 'typeof ConsoleLogger' but required in type 'Car'.
  Overload 3 of 4, '(from: Abstraction<Car>, to: FactoryFunction<Car>): Scope', gave the following error.
    Argument of type 'typeof ConsoleLogger' is not assignable to parameter of type 'FactoryFunction<Car>'.
      Type 'typeof ConsoleLogger' provides no match for the signature '(container: Container): Car'.
test/spec.ts(100,35): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<ConsoleLogger>, to: Newable<ConsoleLogger>): Scope', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'Newable<ConsoleLogger>'.
      Property 'log' is missing in type 'Car' but required in type 'ConsoleLogger'.
  Overload 2 of 4, '(from: Abstraction<ConsoleLogger>, to: ConsoleLogger): void', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'ConsoleLogger'.
      Property 'log' is missing in type 'typeof Car' but required in type 'ConsoleLogger'.
  Overload 3 of 4, '(from: Abstraction<ConsoleLogger>, to: FactoryFunction<ConsoleLogger>): Scope', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'FactoryFunction<ConsoleLogger>'.
      Type 'typeof Car' provides no match for the signature '(container: Container): ConsoleLogger'.
test/spec.ts(103,20): error TS2345: Argument of type 'typeof Logger' is not assignable to parameter of type 'Newable<Logger>'.
  Cannot assign an abstract constructor type to a non-abstract constructor type.
test/spec.ts(104,28): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type 'typeof Logger' is not assignable to parameter of type 'Newable<Logger>'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type 'typeof Logger' is not assignable to parameter of type 'Logger'.
      Property 'log' is missing in type 'typeof Logger' but required in type 'Logger'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Argument of type 'typeof Logger' is not assignable to parameter of type 'FactoryFunction<Logger>'.
      Type 'typeof Logger' provides no match for the signature '(container: Container): Logger'.
test/spec.ts(107,28): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'Newable<Logger>'.
      Property 'log' is missing in type 'Car' but required in type 'Logger'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'Logger'.
      Property 'log' is missing in type 'typeof Car' but required in type 'Logger'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Argument of type 'typeof Car' is not assignable to parameter of type 'FactoryFunction<Logger>'.
      Type 'typeof Car' provides no match for the signature '(container: Container): Logger'.
test/spec.ts(108,28): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type 'typeof BadImplementationLogger' is not assignable to parameter of type 'Newable<Logger>'.
      Construct signature return types 'BadImplementationLogger' and 'Logger' are incompatible.
        The types of 'log' are incompatible between these types.
          Type '(message: number) => void' is not assignable to type '(message: string) => void'.
            Types of parameters 'message' and 'message' are incompatible.
              Type 'string' is not assignable to type 'number'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type 'typeof BadImplementationLogger' is not assignable to parameter of type 'Logger'.
      Property 'log' is missing in type 'typeof BadImplementationLogger' but required in type 'Logger'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Argument of type 'typeof BadImplementationLogger' is not assignable to parameter of type 'FactoryFunction<Logger>'.
      Type 'typeof BadImplementationLogger' provides no match for the signature '(container: Container): Logger'.
test/spec.ts(112,38): error TS2339: Property 'singletonScope' does not exist on type 'void'.
test/spec.ts(113,38): error TS2339: Property 'resolutionScope' does not exist on type 'void'.
test/spec.ts(114,39): error TS2339: Property 'singletonScope' does not exist on type 'void'.
test/spec.ts(115,39): error TS2339: Property 'resolutionScope' does not exist on type 'void'.
test/spec.ts(119,28): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'Newable<Logger>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature 'new (...args: any[]): Logger'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'Logger'.
      Types of property 'log' are incompatible.
        Type '(message: string[]) => void' is not assignable to type '(message: string) => void'.
          Types of parameters 'message' and 'message' are incompatible.
            Type 'string' is not assignable to type 'string[]'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'FactoryFunction<Logger>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature '(container: Container): Logger'.
test/spec.ts(120,35): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<ConsoleLogger>, to: Newable<ConsoleLogger>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'Newable<ConsoleLogger>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature 'new (...args: any[]): ConsoleLogger'.
  Overload 2 of 4, '(from: Abstraction<ConsoleLogger>, to: ConsoleLogger): void', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'ConsoleLogger'.
      Types of property 'log' are incompatible.
        Type '(message: string[]) => void' is not assignable to type '(message: string) => void'.
          Types of parameters 'message' and 'message' are incompatible.
            Type 'string' is not assignable to type 'string[]'.
  Overload 3 of 4, '(from: Abstraction<ConsoleLogger>, to: FactoryFunction<ConsoleLogger>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'FactoryFunction<ConsoleLogger>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature '(container: Container): ConsoleLogger'.
test/spec.ts(121,25): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Car>, to: Newable<Car>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'Newable<Car>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature 'new (...args: any[]): Car'.
  Overload 2 of 4, '(from: Abstraction<Car>, to: Car): void', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'Car'.
      Property 'drive' is missing in type '{ log: (message: string[]) => void; }' but required in type 'Car'.
  Overload 3 of 4, '(from: Abstraction<Car>, to: FactoryFunction<Car>): Scope', gave the following error.
    Argument of type '{ log: (message: string[]) => void; }' is not assignable to parameter of type 'FactoryFunction<Car>'.
      Type '{ log: (message: string[]) => void; }' provides no match for the signature '(container: Container): Car'.
test/spec.ts(124,1): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type '() => Car' is not assignable to parameter of type 'Newable<Logger>'.
      Type '() => Car' provides no match for the signature 'new (...args: any[]): Logger'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type '() => Car' is not assignable to parameter of type 'Logger'.
      Property 'log' is missing in type '() => Car' but required in type 'Logger'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Type 'Car' is not assignable to type 'Logger'.
test/spec.ts(126,28): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<Logger>, to: Newable<Logger>): Scope', gave the following error.
    Argument of type '() => { log: (message: string) => void; } | "this shouldnt happen"' is not assignable to parameter of type 'Newable<Logger>'.
      Type '() => { log: (message: string) => void; } | "this shouldnt happen"' provides no match for the signature 'new (...args: any[]): Logger'.
  Overload 2 of 4, '(from: Abstraction<Logger>, to: Logger): void', gave the following error.
    Argument of type '() => { log: (message: string) => void; } | "this shouldnt happen"' is not assignable to parameter of type 'Logger'.
      Property 'log' is missing in type '() => { log: (message: string) => void; } | "this shouldnt happen"' but required in type 'Logger'.
  Overload 3 of 4, '(from: Abstraction<Logger>, to: FactoryFunction<Logger>): Scope', gave the following error.
    Argument of type '() => { log: (message: string) => void; } | "this shouldnt happen"' is not assignable to parameter of type 'FactoryFunction<Logger>'.
      Type '{ log: (message: string) => void; } | "this shouldnt happen"' is not assignable to type 'Logger'.
        Type '"this shouldnt happen"' is not assignable to type 'Logger'.
test/spec.ts(135,35): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<ConsoleLogger>, to: Newable<ConsoleLogger>): Scope', gave the following error.
    Argument of type '() => { log: (message: string[]) => void; }' is not assignable to parameter of type 'Newable<ConsoleLogger>'.
      Type '() => { log: (message: string[]) => void; }' provides no match for the signature 'new (...args: any[]): ConsoleLogger'.
  Overload 2 of 4, '(from: Abstraction<ConsoleLogger>, to: ConsoleLogger): void', gave the following error.
    Argument of type '() => { log: (message: string[]) => void; }' is not assignable to parameter of type 'ConsoleLogger'.
      Property 'log' is missing in type '() => { log: (message: string[]) => void; }' but required in type 'ConsoleLogger'.
  Overload 3 of 4, '(from: Abstraction<ConsoleLogger>, to: FactoryFunction<ConsoleLogger>): Scope', gave the following error.
    Argument of type '() => { log: (message: string[]) => void; }' is not assignable to parameter of type 'FactoryFunction<ConsoleLogger>'.
      Call signature return types '{ log: (message: string[]) => void; }' and 'ConsoleLogger' are incompatible.
        The types of 'log' are incompatible between these types.
          Type '(message: string[]) => void' is not assignable to type '(message: string) => void'.
            Types of parameters 'message' and 'message' are incompatible.
              Type 'string' is not assignable to type 'string[]'.
test/spec.ts(140,35): error TS2769: No overload matches this call.
  Overload 1 of 4, '(from: Abstraction<ConsoleLogger>, to: Newable<ConsoleLogger>): Scope', gave the following error.
    Argument of type '(c: number) => ConsoleLogger' is not assignable to parameter of type 'Newable<ConsoleLogger>'.
      Type '(c: number) => ConsoleLogger' provides no match for the signature 'new (...args: any[]): ConsoleLogger'.
  Overload 2 of 4, '(from: Abstraction<ConsoleLogger>, to: ConsoleLogger): void', gave the following error.
    Argument of type '(c: number) => ConsoleLogger' is not assignable to parameter of type 'ConsoleLogger'.
      Property 'log' is missing in type '(c: number) => ConsoleLogger' but required in type 'ConsoleLogger'.
  Overload 3 of 4, '(from: Abstraction<ConsoleLogger>, to: FactoryFunction<ConsoleLogger>): Scope', gave the following error.
    Argument of type '(c: number) => ConsoleLogger' is not assignable to parameter of type 'FactoryFunction<ConsoleLogger>'.
      Types of parameters 'c' and 'container' are incompatible.
        Type 'Container' is not assignable to type 'number'.
